Library/OcCryptoLib/Sha2-avx.h

```diff
@@ -0,0 +1,92 @@
+/** @file
+  Copyright (C) 2021, ISP RAS. All rights reserved.
+
+  All rights reserved.
+
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+**/
+
+#ifndef SHA2_AVX_H
+#define SHA2_AVX_H
+
+#include <Library/OcDebugLogLib.h>
+#include <Library/OcCryptoLib.h>
+
+#define UNPACK64(x, str)                         \
+  do {                                           \
+    *((str) + 7) = (UINT8) (x);                  \
+    *((str) + 6) = (UINT8) RShiftU64 ((x),  8);  \
+    *((str) + 5) = (UINT8) RShiftU64 ((x), 16);  \
+    *((str) + 4) = (UINT8) RShiftU64 ((x), 24);  \
+    *((str) + 3) = (UINT8) RShiftU64 ((x), 32);  \
+    *((str) + 2) = (UINT8) RShiftU64 ((x), 40);  \
+    *((str) + 1) = (UINT8) RShiftU64 ((x), 48);  \
+    *((str) + 0) = (UINT8) RShiftU64 ((x), 56);  \
+  } while(0)
+
+#define PACK64(str, x)                           \
+  do {                                           \
+    *(x) =    ((UINT64) *((str) + 7))            \
+           | LShiftU64 (*((str) + 6),  8)        \
+           | LShiftU64 (*((str) + 5), 16)        \
+           | LShiftU64 (*((str) + 4), 24)        \
+           | LShiftU64 (*((str) + 3), 32)        \
+           | LShiftU64 (*((str) + 2), 40)        \
+           | LShiftU64 (*((str) + 1), 48)        \
+           | LShiftU64 (*((str) + 0), 56);       \
+  } while (0)
+
+typedef struct {
+  UINT64 State[8];
+  UINT64 Length;
+  UINT64 TotalLength;
+  UINT8  Block[2 * SHA512_BLOCK_SIZE];
+} SHA512_STATE;
```

Comment:
Could you please make the state common between AVX and non-AVX functions? I guess it will be easier to use this state and update C versions, since they are mostly renames anyway.

The primary reason for this is that I would like Sha384Update to use AVX rather than have separate branching in Sha384. See below.